### PR TITLE
Added jitHelpers for store and load

### PR DESCRIPTION
**Pull request by OussamaSaoudi**
Added jitHelpers for loading and storing flattenable array
element with a slow and fast variant.
-old_fast-jitLoadFlattenableArrayElement: Attempts to use fast
path to load array element. If it fails calls the slow variant.
-old_slow_jitLoadFlattenableArrayElement: Proceeds thorugh slow
path for loading, throwing appropriate exceptions when needed.
-old_fast_jitStoreFlattenableArrayElement: Attempts to use fast
path to store value into array. If it fails, calls the slow\
variant.
-old_slow_jitStoreFlattenableArrayElement: Proceeds through
sllow path for storing, throws appropriate exceptions when
needed.

Added runtime helper to check if storing object in an array
is allowed.
-jitObjectArrayStoreAllowed

Added ValueTypeHelpers loadFlattenableArrayElement and
storeFlattenableArrayElement which perform aaload and
aastore functions respectively. Both handle flattened and
unflattened cases. Both contain an option for either a fast
or a slow variant.
-loadFlattenableArrayElement: Loads flattenable array element
-storeFlattenableArrayElement: Stores flattenable array element
-loadFlattenableArrayElement: c function that initializes
required objects and calls loadFlattenableArrayElement
-storeFlattenableArrayElement: c function that initializes
required objects and calls storeFlattenableArrayElement

Replaced existing code in aaload and aastore functions in
ByteCodeInterpreter.hpp with calls to loadFlattenableArrayElement
and storeFlattenableArrayElement. Functionality remains the same.

Related to: https://github.com/eclipse/openj9/issues/9627

Signed-off-by: Oussama Saoudi <oussama.saoudi@ibm.com>

### CODE REVIEW

**Comment by OussamaSaoudi** [comment]
@tajila

File: runtime/vm/ValueTypeHelpers.cpp [hunk 1]
```diff
@@ -69,4 +69,85 @@ isClassRefQtype(J9Class *cpContextClass, U_16 cpIndex)
...
+ * Returns null if newObjectRef retrieval fails.
+ *
+ * @param[in] currentThread thread token
+ * @param[in] receiverObject arrayobject
+ * @param[in] index array index
+ *
+ * @return array element
+ */
+j9_object_t
+jitAaload(J9VMThread currentThread, j9_object_t *arrayObject, U_32 index)
```
Comment (tajila) [inline runtime/vm/ValueTypeHelpers.cpp:128]: add a new bool param called fast, if its fast you can use the fast allocator, otherwise the slow one.

File: runtime/codert_vm/cnathelp.cpp [hunk 2]
```diff
@@ -748,7 +748,90 @@ old_slow_jitLoadFlattenableArrayElement(J9VMThread *currentThread)
...
+	}
+	U_32 arrayLength = J9INDEXABLEOBJECT_SIZE(_currentThread, arrayObject);
+	if (index >= arrayLength) {
+		goto slow;
+	}
+	j9_object_t value = jitAaload(currentThread, arrayObject, index);
+	if (J9_UNEXPECTED(NULL == value)) {
+			goto slow;
+	} else {
+		*(j9object_t*)_sp = value;
```
Comment (tajila) [inline runtime/codert_vm/cnathelp.cpp:827]: no _sp in this context, need to return a value with `JIT_RETURN_UDATA(obj);`

File: runtime/codert_vm/cnathelp.cpp [hunk 3]
```diff
@@ -748,7 +748,90 @@ old_slow_jitLoadFlattenableArrayElement(J9VMThread *currentThread)
...
+				}
+
+				_objectAccessBarrier.copyObjectFieldsFromFlattenedArrayElement(currentThread, (J9ArrayClass *) arrayrefClass, newObjectRef, (J9IndexableObject *) arrayref, index);
+				value = newObjectRef;
+			} else
+#endif /* if defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
+			{
+				value = _objectAccessBarrier.inlineIndexableObjectReadObject(currentThread, arrayref, index);
+			}
+			*(j9object_t*)_sp = value;
```
Comment (tajila) [inline runtime/codert_vm/cnathelp.cpp:806]: no _sp in this context, need to return a value with JIT_RETURN_UDATA(obj);

File: runtime/codert_vm/cnathelp.cpp [hunk 4]
```diff
@@ -748,7 +748,90 @@ old_slow_jitLoadFlattenableArrayElement(J9VMThread *currentThread)
 void* J9FASTCALL
 old_fast_jitLoadFlattenableArrayElement(J9VMThread *currentThread)
 {
-	return NULL;
+	OLD_JIT_HELPER_PROLOGUE(1);
+	DECLARE_JIT_PARM(j9_object_t *, arrayObject, 1);
+	DECLARE_JIT_INT_PARM(index, 1);
+	void *slowPath = NULL;
+	if (fast_jitLoadFlattenableArrayElementImpl(currentThread, arrayObject, index)) {
```
Comment (tajila) [inline runtime/codert_vm/cnathelp.cpp:755]: dont think this helper is needed, you have just put the code in here

File: runtime/vm/ValueTypeHelpers.cpp [hunk 5]
```diff
@@ -69,4 +69,85 @@ isClassRefQtype(J9Class *cpContextClass, U_16 cpIndex)
...
+ * doesnt need to throw.
+ *
+ * @param[in] currentThread thread token
+ * @param[in] receiverObject arrayobject
+ * @param[in] index array index
+ * @param[in] paramObject obj arg
+ *
+ */
+void
+jitAastore(J9VMThread currentThread, j9_object_t *arrayObject, U_32 index, j9_object_t *paramObject)
```
Comment (tajila) [inline runtime/vm/ValueTypeHelpers.cpp:89]: this will have to be added to a function table, same as the other one. you can rename them to [load/store]FlattenableArrayElement

File: runtime/codert_vm/cnathelp.cpp [hunk 6]
```diff
@@ -748,7 +748,90 @@ old_slow_jitLoadFlattenableArrayElement(J9VMThread *currentThread)
...
+		slowPath = (void*) old_slow_jitLoadFlattenableArrayElement;
+	}
+	return slowPath;
+}
+
+void* J9FASTCALL
+old_slow_jitLoadFlattenableArrayElementImpl(J9VMThread *currentThread)
+{
+	SLOW_JIT_HELPER_PROLOGUE();
+	J9Class *arrayref = (J9Class*)currentThread->floatTemp1;
```
Comment (tajila) [inline runtime/codert_vm/cnathelp.cpp:765]: you need the index as well

File: runtime/codert_vm/cnathelp.cpp [hunk 7]
```diff
@@ -748,7 +748,90 @@ old_slow_jitLoadFlattenableArrayElement(J9VMThread *currentThread)
...
+		U_32 arrayLength = J9INDEXABLEOBJECT_SIZE(currentThread, arrayref);
+		U_32 index = *(U_32*)_sp;
+		/* By using U_32 for index, we also catch the negative case, as all negative values are
+			* greater than the maximum array size (31 bits unsigned).
+			*/
+		if (index >= arrayLength) {
+			currentThread->tempSlot = (UDATA)index;
+			rc = THROW_AIOB;
+		} else {
+			j9object_t value = NULL;
```
Comment (tajila) [inline runtime/codert_vm/cnathelp.cpp:778]: should  be using the VT helper here as well, with fast = FALSE

File: runtime/codert_vm/cnathelp.cpp [hunk 8]
```diff
@@ -748,7 +748,90 @@ old_slow_jitLoadFlattenableArrayElement(J9VMThread *currentThread)
...
+	return slowPath;
+}
+
+void* J9FASTCALL
+old_slow_jitLoadFlattenableArrayElementImpl(J9VMThread *currentThread)
+{
+	SLOW_JIT_HELPER_PROLOGUE();
+	J9Class *arrayref = (J9Class*)currentThread->floatTemp1;
+	if (NULL == arrayref) {
+		rc = THROW_NPE;
```
Comment (tajila) [inline runtime/codert_vm/cnathelp.cpp:767]: when throwing exceptions you need to set it to the correct exception routine. Examples are:

`addr = setHeapOutOfMemoryErrorFromJIT(currentThread);`
`addr = setCurrentExceptionFromJIT(currentThread, J9VMCONSTANTPOOL_JAVALANGINSTANTIATIONERROR | J9_EX_CTOR_CLASS, J9VM_J9CLASS_TO_HEAPCLASS(objectClass));`

File: runtime/codert_vm/cnathelp.cpp [hunk 9]
```diff
@@ -742,13 +742,68 @@ old_fast_jitPutFlattenableStaticField(J9VMThread *currentThread)
...
+	if (index >= arrayLength) {
+		goto slow;
+	}
+	value = (j9object_t) currentThread->javaVM->internalVMFunctions->loadFlattenableArrayElement(currentThread, arrayObject, index, true);
+	if (J9_UNEXPECTED(NULL == value)) {
+			goto slow;
+	}
+
+done:
+	if (!slowPathUsed) {
```
Comment (tajila) [inline runtime/codert_vm/cnathelp.cpp:797]: this can be moved above the `done:` label so you dont need the `!slowPathUsed`

File: runtime/oti/j9nonbuilder.h [hunk 10]
```diff
@@ -3496,6 +3496,8 @@ typedef struct J9ClassCastParms {
 
 typedef struct J9JITConfig {
 	IDATA  ( *entryPoint)(struct J9JITConfig *jitConfig, struct J9VMThread *vmStruct, J9Method *method, void *oldStartPC) ;
+	void *old_fast_jitLoadFlattenableArrayElement;
```
Comment (tajila) [inline runtime/oti/j9nonbuilder.h:3499]: why is this moved?

File: runtime/oti/j9nonbuilder.h [hunk 11]
```diff
@@ -3496,6 +3496,8 @@ typedef struct J9ClassCastParms {
...
+	void *old_slow_jitLoadFlattenableArrayElement;
```
Comment (tajila) [inline runtime/oti/j9nonbuilder.h:3500]: i dont think you need the slow version

File: runtime/vm/ValueTypeHelpers.cpp [hunk 12]
```diff
@@ -168,5 +168,20 @@ arrayElementSize(J9ArrayClass* arrayClass)
...
+{
+        MM_ObjectAccessBarrierAPI objectAccessBarrier(currentThread);
+        VM_ValueTypeHelpers::jitAastore(currentThread, objectAccessBarrier, arrayObject, index, paramObject, fast);
+
+}
+
+j9object_t
+loadFlattenableArrayElement(J9VMThread *currentThread, j9object_t arrayObject, U_32 index, BOOLEAN fast)
+{
+        MM_ObjectAccessBarrierAPI objectAccessBarrier(currentThread);
```
Comment (tajila) [inline runtime/vm/ValueTypeHelpers.cpp:182]: indenting needs to be fixed, make sure you are using tabs not spaces

File: runtime/vm/ValueTypeHelpers.hpp [hunk 13]
```diff
@@ -255,6 +255,75 @@ class VM_ValueTypeHelpers {
 		return rc;
 	}
 
+	static VMINLINE void
+	jitAastore(J9VMThread *currentThread, MM_ObjectAccessBarrierAPI _objectAccessBarrier, j9object_t arrayObject, U_32 index, j9object_t *paramObject, bool fast)
+	{
+	#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
```
Comment (tajila) [inline runtime/vm/ValueTypeHelpers.hpp:261]: same thing here with the indenting

File: runtime/vm/ValueTypeHelpers.hpp [hunk 14]
```diff
@@ -255,6 +255,75 @@ class VM_ValueTypeHelpers {
...
+			if (index >= arrayLength) {
+				currentThread->tempSlot = (UDATA)index;
+				goto done;
+			} else {
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+				J9Class *arrayrefClass = J9OBJECT_CLAZZ(currentThread, arrayObject);
+				if (J9_IS_J9CLASS_FLATTENED(arrayrefClass)) {
+					j9object_t newObjectRef = _objectAllocate.inlineAllocateObject(currentThread, ((J9ArrayClass*)arrayrefClass)->leafComponentType, false, false);
+					if (NULL == newObjectRef) {
+						newObjectRef = currentThread->javaVM->memoryManagerFunctions->J9AllocateObject(currentThread, ((J9ArrayClass*)arrayrefClass)->leafComponentType, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
```
Comment (tajila) [inline runtime/vm/ValueTypeHelpers.hpp:308]: if you are doing an out-of-line allocation youll need to build a special frame in case of a GC

Comment (OussamaSaoudi) [reply runtime/vm/ValueTypeHelpers.hpp:310]: Could you take a look at what I've got down now to make sure it's correct?

File: runtime/codert_vm/cnathelp.cpp [hunk 15]
```diff
@@ -742,13 +742,64 @@ old_fast_jitPutFlattenableStaticField(J9VMThread *currentThread)
 void* J9FASTCALL
 old_slow_jitLoadFlattenableArrayElement(J9VMThread *currentThread)
 {
-	return NULL;
+	SLOW_JIT_HELPER_PROLOGUE();
+	j9object_t arrayObject = (j9object_t)currentThread->floatTemp1;
+	U_32 index = *((U_32 *)currentThread->floatTemp2);
+	void *addr = NULL;
+	if (NULL == arrayObject) {
```
Comment (tajila) [inline runtime/codert_vm/cnathelp.cpp:749]: need frame build before setting the exception

File: runtime/codert_vm/cnathelp.cpp [hunk 16]
```diff
@@ -742,13 +742,64 @@ old_fast_jitPutFlattenableStaticField(J9VMThread *currentThread)
...
+		addr = setCurrentExceptionFromJIT(currentThread, J9VMCONSTANTPOOL_JAVALANGNULLPOINTEREXCEPTION, NULL);
+	} else {
+		j9object_t value = NULL;
+		U_32 arrayLength = J9INDEXABLEOBJECT_SIZE(_currentThread, arrayObject);
+		if (index >= arrayLength) {
+			currentThread->tempSlot = (UDATA)index;
+			addr = setCurrentExceptionFromJIT(currentThread, J9VMCONSTANTPOOL_JAVALANGARRAYINDEXOUTOFBOUNDSEXCEPTION, NULL);
```
Comment (tajila) [inline runtime/codert_vm/cnathelp.cpp:756]: same here, and below

File: runtime/vm/ValueTypeHelpers.cpp [hunk 17]
```diff
@@ -168,5 +168,20 @@ arrayElementSize(J9ArrayClass* arrayClass)
         Assert_VM_notNull(arrayClass);
         return arrayClass->flattenedElementSize;
 }
+void
+storeFlattenableArrayElement(J9VMThread *currentThread, j9object_t arrayObject, U_32 index, j9object_t *paramObject, BOOLEAN fast)
+{
+        MM_ObjectAccessBarrierAPI objectAccessBarrier(currentThread);
+        VM_ValueTypeHelpers::jitAastore(currentThread, objectAccessBarrier, arrayObject, index, paramObject, fast);
+
```
Comment (tajila) [inline runtime/vm/ValueTypeHelpers.cpp:176]: extra line here

File: runtime/vm/ValueTypeHelpers.cpp [hunk 18]
```diff
@@ -168,5 +168,20 @@ arrayElementSize(J9ArrayClass* arrayClass)
...
+}
+
+j9object_t
+loadFlattenableArrayElement(J9VMThread *currentThread, j9object_t arrayObject, U_32 index, BOOLEAN fast)
+{
+        MM_ObjectAccessBarrierAPI objectAccessBarrier(currentThread);
+        MM_ObjectAllocationAPI objectAllocate(currentThread);
+        return VM_ValueTypeHelpers::jitAaload(currentThread, objectAccessBarrier, objectAllocate, arrayObject, index, fast);
```
Comment (tajila) [inline runtime/vm/ValueTypeHelpers.cpp:184]: that boolean `fast` wont compile on windows, use `fast != false` instead

File: runtime/vm/ValueTypeHelpers.hpp [hunk 19]
```diff
@@ -255,6 +255,82 @@ class VM_ValueTypeHelpers {
...
+				currentThread->tempSlot = (UDATA)index;
+				goto done;
+			} else {
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+				J9Class *arrayrefClass = J9OBJECT_CLAZZ(currentThread, arrayObject);
+				if (J9_IS_J9CLASS_FLATTENED(arrayrefClass)) {
+					j9object_t newObjectRef = _objectAllocate.inlineAllocateObject(currentThread, ((J9ArrayClass*)arrayrefClass)->leafComponentType, false, false);
+					if (NULL == newObjectRef) {
+						UDATA *bp = VM_OutOfLineINL_Helpers::buildSpecialStackFrame(currentThread, J9SF_FRAME_TYPE_GENERIC_SPECIAL, 0, false);
+						VM_VMHelpers::pushObjectInSpecialFrame(currentThread, arrayObject);
```
Comment (tajila) [inline runtime/vm/ValueTypeHelpers.hpp:309]: instead of pushing frame here, just make sure caller pushes frame if fast is set to false. You can add something in the comments

File: runtime/vm/ValueTypeHelpers.hpp [hunk 20]
```diff
@@ -255,6 +255,82 @@ class VM_ValueTypeHelpers {
...
+					goto done;
+				}
+				_objectAccessBarrier.copyObjectFieldsFromFlattenedArrayElement(currentThread, (J9ArrayClass *) arrayrefClass, newObjectRef, (J9IndexableObject *)arrayObject, index);
+				value = newObjectRef;
+			} else
+#endif /* if defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
+			{
+				value = _objectAccessBarrier.inlineIndexableObjectReadObject(currentThread, arrayObject, index);
+			}
+		} else if (NULL != arrayObject) {
```
Comment (tajila) [inline runtime/vm/ValueTypeHelpers.hpp:297]: this else branch is not needed, all the checks are done before calling this function

File: runtime/vm/ValueTypeHelpers.hpp [hunk 21]
```diff
@@ -255,6 +255,82 @@ class VM_ValueTypeHelpers {
...
+	static VMINLINE j9object_t
+	jitAaload(J9VMThread *currentThread, MM_ObjectAccessBarrierAPI _objectAccessBarrier, MM_ObjectAllocationAPI _objectAllocate, j9object_t arrayObject, U_32 index, bool fast)
+	{
+		j9object_t value = NULL;
+
+		if (fast) {
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+			J9Class *arrayrefClass = J9OBJECT_CLAZZ(currentThread, arrayObject);
+			if (J9_IS_J9CLASS_FLATTENED(arrayrefClass)) {
+				j9object_t newObjectRef = _objectAllocate.inlineAllocateObject(currentThread, ((J9ArrayClass*)arrayrefClass)->leafComponentType, false, false);
```
Comment (tajila) [inline runtime/vm/ValueTypeHelpers.hpp:285]: fast should be checked here, if fast use `_objectAllocate.inlineAllocateObject` else use `currentThread->javaVM->memoryManagerFunctions->J9AllocateObject`

File: runtime/vm/ValueTypeHelpers.hpp [hunk 22]
```diff
@@ -255,6 +255,82 @@ class VM_ValueTypeHelpers {
 		return rc;
 	}
 
+	static VMINLINE void
```
Comment (tajila) [inline runtime/vm/ValueTypeHelpers.hpp:258]: add javadocs here as well

File: runtime/vm/ValueTypeHelpers.hpp [hunk 23]
```diff
@@ -255,6 +255,82 @@ class VM_ValueTypeHelpers {
...
+	jitAastore(J9VMThread *currentThread, MM_ObjectAccessBarrierAPI _objectAccessBarrier, j9object_t arrayObject, U_32 index, j9object_t *paramObject, bool fast)
```
Comment (tajila) [inline runtime/vm/ValueTypeHelpers.hpp:259]: rename to `store/loadFlattenableArrayElement`

File: runtime/vm/ValueTypeHelpers.hpp [hunk 24]
```diff
@@ -255,6 +255,82 @@ class VM_ValueTypeHelpers {
...
+#endif /* if defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
+				{
+					value = _objectAccessBarrier.inlineIndexableObjectReadObject(currentThread, arrayObject, index);
+				}
+			}
+		}
+done:
+		return value;
+	}
+
```
Comment (tajila) [inline runtime/vm/ValueTypeHelpers.hpp:333]: these functions should be used in the bytecodeinterpreter so we dont duplicate logic

File: runtime/vm/ValueTypeHelpers.hpp [hunk 25]
```diff
@@ -417,6 +417,81 @@ class VM_ValueTypeHelpers {
 		}
 	}
 
+	/**
+	 * Performs an aastore operation on an object. Handles flattened and non-flattened cases.
+	 *
+	 * Assumes recieverObject and paramObject are not null.
```
Comment (OussamaSaoudi) [inline runtime/vm/ValueTypeHelpers.hpp:423]: Should I add that it assumes the index is less than the array size and that it does not throw exceptions?

File: runtime/codert_vm/cnathelp.cpp [hunk 26]
```diff
@@ -885,13 +885,69 @@ old_fast_jitPutFlattenableStaticField(J9VMThread *currentThread)
...
+	if (NULL == arrayObject) {
+		buildJITResolveFrameForRuntimeHelper(currentThread, parmCount);
+		addr = setCurrentExceptionFromJIT(currentThread, J9VMCONSTANTPOOL_JAVALANGNULLPOINTEREXCEPTION, NULL);
+	} else {
+		j9object_t value = NULL;
+		U_32 arrayLength = J9INDEXABLEOBJECT_SIZE(_currentThread, arrayObject);
+		if (index >= arrayLength) {
+			currentThread->tempSlot = (UDATA)index;
+			buildJITResolveFrameForRuntimeHelper(currentThread, parmCount);
+			addr = setCurrentExceptionFromJIT(currentThread, J9VMCONSTANTPOOL_JAVALANGARRAYINDEXOUTOFBOUNDSEXCEPTION, NULL);
```
Comment (OussamaSaoudi) [inline runtime/codert_vm/cnathelp.cpp:901]: This line will then go to JIT_RETURN_UDATA(value), returning NULL. I want to confirm that this is fine?

File: runtime/codert_vm/cnathelp.cpp [hunk 27]
```diff
@@ -885,13 +885,69 @@ old_fast_jitPutFlattenableStaticField(J9VMThread *currentThread)
...
+		} else {
+			buildJITResolveFrame(currentThread, J9_SSF_JIT_RESOLVE, parmCount);
+			PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, arrayObject);
+			value = currentThread->javaVM->internalVMFunctions->loadFlattenableArrayElement(currentThread, arrayObject, index, false);
```
Comment (OussamaSaoudi) [inline runtime/codert_vm/cnathelp.cpp:905]: fast == false, so I need to build a special frame. Is this the correct way to do it in this context, or should I use OutOfLineNL?

File: runtime/vm/ValueTypeHelpers.hpp (hunk 24 above)
Comment (OussamaSaoudi) [reply runtime/vm/ValueTypeHelpers.hpp:494]: To confirm, I should replace the code currently in ByteCodeInterpreter with these functions?

Comment (tajila) [reply runtime/vm/ValueTypeHelpers.hpp:494]: yes

File: runtime/vm/ValueTypeHelpers.hpp (hunk 25 above)
Comment (tajila) [reply runtime/vm/ValueTypeHelpers.hpp:423]: yes

Comment (tajila) [reply runtime/vm/ValueTypeHelpers.hpp:423]: you can add that all AIOB exceptions must be thrown before hand

File: runtime/codert_vm/cnathelp.cpp (hunk 27 above)
Comment (tajila) [reply runtime/codert_vm/cnathelp.cpp:905]: the frame should be built before call the function

Comment (OussamaSaoudi) [reply runtime/codert_vm/cnathelp.cpp:905]: So I build it regardless of the path being fast or slow?

File: runtime/codert_vm/cnathelp.cpp (hunk 26 above)
Comment (tajila) [reply runtime/codert_vm/cnathelp.cpp:901]: if you are throwing an exception you dont need to set the return val, you can just jum straight to done

Comment (tajila) [reply runtime/codert_vm/cnathelp.cpp:901]: or conversely, just move up the `JIT_RETURN_UDATA(value);` to after the NULL check for the return value

File: runtime/vm/BytecodeInterpreter.hpp [hunk 28]
```diff
@@ -5933,21 +5924,12 @@ class INTERPRETER_CLASS
...
-						if (J9_IS_J9CLASS_FLATTENED(arrayrefClass)) {
-							_objectAccessBarrier.copyObjectFieldsToFlattenedArrayElement(_currentThread, arrayrefClass, value, (J9IndexableObject *) arrayref, index);
-						} else {
-							_objectAccessBarrier.inlineIndexableObjectStoreObject(_currentThread, arrayref, index, value);
-						}
-					} else
-#endif /* if defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
-					{
-						_objectAccessBarrier.inlineIndexableObjectStoreObject(_currentThread, arrayref, index, value);
+					if (J9_IS_J9CLASS_VALUETYPE(arrayrefClass->componentType) && NULL == value) {
```
Comment (tajila) [inline runtime/vm/BytecodeInterpreter.hpp:5927]: add brackets `(NULL == value)`

File: runtime/vm/BytecodeInterpreter.hpp [hunk 29]
```diff
@@ -5872,34 +5872,25 @@ class INTERPRETER_CLASS
...
-						newObjectRef = _vm->memoryManagerFunctions->J9AllocateObject(_currentThread, ((J9ArrayClass*)arrayrefClass)->leafComponentType, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
-						VMStructHasBeenUpdated(REGISTER_ARGS);
-						arrayref = popObjectInSpecialFrame(REGISTER_ARGS);
-						restoreGenericSpecialStackFrame(REGISTER_ARGS);
-						if (J9_UNEXPECTED(NULL == newObjectRef)) {
-							rc = THROW_HEAP_OOM;
-							return rc;
-						}
-						arrayrefClass = VM_VMHelpers::currentClass(arrayrefClass);
+				if (NULL == value && J9_IS_J9CLASS_FLATTENED(arrayrefClass)) {
```
[the remark this answers is not in the document]
Comment (OussamaSaoudi) [reply runtime/vm/BytecodeInterpreter.hpp:5878]: Removing the `J9_IS_J9CLASS_FLATTENED` will cause a system dump. I believe the reason is `value = _objectAccessBarrier.inlineIndexableObjectReadObject(currentThread, receiverObject, index);` is being run in all cases in which the value is not flattened. In the case that it does return null it will always hit the HEAP_OOM error (which should occur only if the slow path is used and returns null, not the non flattened path returns null) This was the simplest way I could come up with to prevent that.

[the remark this answers is not in the document]
Comment (tajila) [reply runtime/vm/BytecodeInterpreter.hpp:5878]: The problem is that you have one path for value types and non-valuetypes. Basically you need a
```
#ifdef valuetype
if (J9_IS_ValueType(arrayrefClass)) {
//your code here
} else
#endif /* valuetype */
{

}

File: runtime/vm/ValueTypeHelpers.hpp [hunk 30]
```diff
@@ -417,6 +417,87 @@ class VM_ValueTypeHelpers {
...
+	 * @param[in] receiverObject arrayObject
+	 * @param[in] index array index
+	 * @param[in] paramObject obj arg
+	 */
+	static VMINLINE void
+	storeFlattenableArrayElement(J9VMThread *currentThread, MM_ObjectAccessBarrierAPI _objectAccessBarrier, j9object_t arrayRef, U_32 index, j9object_t value)
+	{
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+		J9ArrayClass *arrayrefClass = (J9ArrayClass *) J9OBJECT_CLAZZ(currentThread, arrayRef);
+		if (J9_IS_J9CLASS_VALUETYPE(arrayrefClass->componentType)) {
```
Comment (tajila) [inline runtime/vm/ValueTypeHelpers.hpp:437]: you dont need the vt check here

Comment (tajila) [reply runtime/vm/ValueTypeHelpers.hpp:437]: the flattened check is sufficient

File: runtime/codert_vm/cnathelp.cpp [hunk 31]
```diff
@@ -885,13 +885,73 @@ old_fast_jitPutFlattenableStaticField(J9VMThread *currentThread)
...
+	void *addr = NULL;
+	void *oldPC = currentThread->jitReturnAddress;
+	if (NULL == arrayObject) {
+		buildJITResolveFrameForRuntimeHelper(currentThread, parmCount);
+		addr = setCurrentExceptionFromJIT(currentThread, J9VMCONSTANTPOOL_JAVALANGNULLPOINTEREXCEPTION, NULL);
+	} else {
+		j9object_t value = NULL;
+		U_32 arrayLength = J9INDEXABLEOBJECT_SIZE(_currentThread, arrayObject);
+		if (index >= arrayLength) {
+			currentThread->tempSlot = (UDATA)index;
```
[the remark this answers is not in the document]
Comment (OussamaSaoudi) [reply runtime/codert_vm/cnathelp.cpp:900]: This is what is done on the ByteCodeInterpreter version of aaload  if the index is greater than the array length. Let me know if you think it is not needed.

[the remark this answers is not in the document]
Comment (tajila) [reply runtime/codert_vm/cnathelp.cpp:900]: In the interpreter we have a special routine that will look at the tempSlot so that it can print a more helpful exception message. We dont have that here.

File: runtime/codert_vm/cnathelp.cpp [hunk 32]
```diff
@@ -213,6 +213,24 @@ restoreBranchJITResolveFrame(J9VMThread *currentThread)
 	restoreJITResolveFrame(currentThread, NULL, false, false);
 }
 
+VMINLINE bool
```
Comment (tajila) [inline runtime/codert_vm/cnathelp.cpp:216]: you dont need to copy this function, I didnt realize this is what you were talking about. Yes, your original suggestion of moving this into VMHelpers.hpp is correct. This way it can be shared by the interpreter and the runtime helpres

File: runtime/codert_vm/cnathelp.cpp [hunk 33]
```diff
@@ -885,25 +903,150 @@ old_fast_jitPutFlattenableStaticField(J9VMThread *currentThread)
...
+			} else {
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+				J9ArrayClass *arrayrefClass = (J9ArrayClass *) J9OBJECT_CLAZZ(currentThread, arrayref);
+				if (J9_IS_J9CLASS_VALUETYPE(arrayrefClass->componentType) && (NULL == value)) {
+					buildJITResolveFrameForRuntimeHelper(currentThread, parmCount);
+					addr = setCurrentExceptionFromJIT(currentThread, J9VMCONSTANTPOOL_JAVALANGNULLPOINTEREXCEPTION, NULL);;
+				} else
+#endif /* if defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
+				{
+					currentThread->javaVM->internalVMFunctions->storeFlattenableArrayElement(currentThread, arrayref, index, value);
```
Comment (tajila) [inline runtime/codert_vm/cnathelp.cpp:1002]: its not actually possible to get here, you can only be in the slow path if you are throwing an exception, since the storing the element cant fail

File: runtime/vm/BytecodeInterpreter.hpp [hunk 34]
```diff
@@ -5875,34 +5875,23 @@ class INTERPRETER_CLASS
...
-						newObjectRef = _vm->memoryManagerFunctions->J9AllocateObject(_currentThread, ((J9ArrayClass*)arrayrefClass)->leafComponentType, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
-						VMStructHasBeenUpdated(REGISTER_ARGS);
-						arrayref = popObjectInSpecialFrame(REGISTER_ARGS);
-						restoreGenericSpecialStackFrame(REGISTER_ARGS);
-						if (J9_UNEXPECTED(NULL == newObjectRef)) {
-							rc = THROW_HEAP_OOM;
-							return rc;
-						}
-						arrayrefClass = VM_VMHelpers::currentClass(arrayrefClass);
+				if ((NULL == value) && J9_IS_J9CLASS_VALUETYPE(arrayrefClass)) {
```
Comment (tajila) [inline runtime/vm/BytecodeInterpreter.hpp:5881]: add a comment saying `"We only get here due to an allocation failure"`

File: runtime/codert_vm/cnathelp.cpp [hunk 35]
```diff
@@ -885,25 +885,139 @@ old_fast_jitPutFlattenableStaticField(J9VMThread *currentThread)
...
 old_slow_jitStoreFlattenableArrayElement(J9VMThread *currentThread)
 {
-	return NULL;
+	SLOW_JIT_HELPER_PROLOGUE();
+	j9object_t arrayref = (j9object_t)currentThread->floatTemp1;
+	j9object_t value = (j9object_t)currentThread->floatTemp2;
+	U_32 index = *(U_32 *)&currentThread->floatTemp3;
+	U_32 arrayLength = 0;
+	void *addr = NULL;
+	if (NULL == arrayref) {
```
Comment (tajila) [inline runtime/codert_vm/cnathelp.cpp:963]: since this code has to throw, you can just build once the frame upfront

File: runtime/vm/ValueTypeHelpers.hpp [hunk 36]
```diff
@@ -417,6 +417,83 @@ class VM_ValueTypeHelpers {
...
+		j9object_t newObjectRef = NULL;
+		J9Class *arrayrefClass = J9OBJECT_CLAZZ(currentThread, receiverObject);
+		if (J9_IS_J9CLASS_FLATTENED(arrayrefClass)) {
+			if (fast) {
+				newObjectRef = _objectAllocate.inlineAllocateObject(currentThread, ((J9ArrayClass*)arrayrefClass)->leafComponentType, false, false);
+				if(J9_UNEXPECTED(NULL == newObjectRef)) {
+					goto done;
+				}
+			} else {
+				newObjectRef = currentThread->javaVM->memoryManagerFunctions->J9AllocateObject(currentThread, ((J9ArrayClass*)arrayrefClass)->leafComponentType, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
```
[the remark this answers is not in the document]
Comment (OussamaSaoudi) [reply runtime/vm/ValueTypeHelpers.hpp:478]: It's stated in the function's comment 
> "If fast == false, special stack frame must be built and receiverObject must be pushed onto it."
Let me know if you'd still like me to change it.

[the remark this answers is not in the document]
Comment (tajila) [reply runtime/vm/ValueTypeHelpers.hpp:478]: ah forgot about that. ignore that comment

**Review by tajila — APPROVED** [review]

**Comment by tajila** [comment]
Jenkins test sanity,extended xlinuxval jdknext

**Comment by tajila** [comment]
Jenkins test sanity win jdk8

**Comment by tajila** [comment]
Compilation failures:
```
21:37:37  ut_j9codertvm.c
21:37:37  cnathelp.cpp(936) : error C2220: warning treated as error - no 'object' file generated
21:37:37  cnathelp.cpp(936) : warning C4018: '>=' : signed/unsigned mismatch
21:37:37  cnathelp.cpp(998) : warning C4018: '>=' : signed/unsigned mismatch
```
@OussamaSaoudi

**Comment by tajila** [comment]
@OussamaSaoudi have the compilation failures been fixed?

File: runtime/vm/BytecodeInterpreter.hpp [hunk 37]
```diff
@@ -5875,34 +5857,24 @@ class INTERPRETER_CLASS
...
```
Comment (tajila) [inline runtime/vm/BytecodeInterpreter.hpp:5863]: This is the problem, `J9_IS_J9CLASS_VALUETYPE(arrayrefClass)`, arrayclasses are never marked as VT's only the element types. This should remain as the `J9_IS_J9CLASS_FLATTENED` test

**Comment by tajila** [comment]
Jenkins test sanity,extended xlinuxval jdknext

**Comment by tajila** [comment]
Jenkins test sanity win jdk8

**Comment by tajila** [comment]
Jenkins test sanity,extended xlinuxval jdknext
